Bio/PDB/PDBServer.py

```diff
@@ -0,0 +1,173 @@
+
+SERVER_REGEX = re.compile(
+    r"^(?P<protocol>(http|ftp)s?):\/{2}(ftp\.)?(?P<domain>.*\..*?)(\/)?$"
+)
+
+
+@dataclasses.dataclass(frozen=True)
+class PDBServer:
+    """PDB server immutable object."""
+
+    domain: str
+    label: str
+    pdb_directory: str = "/pub/pdb"
+
+    @property
+    def ftp_subdomain(self):
+        """PDB Server ftp sub domain."""
+        return f"ftp.{self.domain}"
+
+    def __str__(self) -> str:
+        return f"{self.label} ({self.domain})"
+
+
+PDB_SERVERS = [
+    PDBServer(domain="wwpdb.org", label="Worldwide"),
+    PDBServer(domain="rcsb.org", label="United States"),
+    PDBServer(
+        domain="ebi.ac.uk", label="United Kingdom", pdb_directory="/pub/databases/pdb"
+    ),
+    PDBServer(domain="pdbj.org", label="Japan"),
+]
+
+
+class PDBServerProtocol(enum.IntEnum):
+    """Available protocol to access to PDB servers."""
+
+    HTTPS = 443
+    FTP = 21
+
+    @property
+    def port(self) -> int:
+        """Protocol port."""
+        return self.value
+
+    def __str__(self) -> str:
+        return f"{self.name} ({self.value})"
+
+
+class PDBServersConnectionError(Exception):
+    """Unable to connect to any PDB server."""
+
+    def __init__(self, protocol: PDBServerProtocol) -> None:
+        """Build PDBServersConnectionError message."""
+        super().__init__(
+            f"Unable to connect to any PDB servers (protocol: {protocol.name}, servers: {', '.join([str(server) for server in PDB_SERVERS])})."
+        )
+
+
+class UnsupportedServerError(Exception):
+    """PDB Server is not supported."""
+
+    def __init__(self, server: str) -> None:
+        """Build UnsupportedServerError message."""
+        super().__init__(
+            f"PDB server is not supported (server: {server}, supported servers: {', '.join([str(server) for server in PDB_SERVERS])})."
+        )
+
+
+class UnsupportedProtocolError(Exception):
+    """Protocol is not supported."""
+
+    def __init__(self, protocol: str) -> None:
+        """Build UnsupportedProtocolError message."""
+        super().__init__(
+            f"Protocol is not supported (protocol: {protocol}, supported protocol: {', '.join([str(_protocol) for _protocol in PDBServerProtocol])})."
+        )
+
+
+@functools.cache
+def get_server_connection_timing(
+    server: PDBServer, protocol: PDBServerProtocol
+) -> int | None:
+    """Duration to connect to a PDB server."""
+    domain = (
+        server.ftp_subdomain if protocol == PDBServerProtocol.FTP else server.domain
+    )
+    with _build_socket() as server_socket:
+        time_before_connect = time.time()
+        try:
+            server_socket.connect((domain, protocol.port))
+        except Exception:
+            return None
+        return time.time() - time_before_connect
+
+
+@contextlib.contextmanager
+def _build_socket():
+    """Context manager auto closing socket."""
+    server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    server_socket.settimeout(1)
```

Comment:
Could be a little bit short, but checking all the servers could be long if the timeout is too important.

it seems to be a good compromise to me, but maybe we can introduce an environment variable to make it customizable.